Feature/buff window

EvgenijK opened the pull request:
It would be nice to have indicators for some timed effects on character. Especially for resistances.
There is even a 4-year old topic on forum about it https://tomenet.eu/phpBB3/viewtopic.php?f=7&t=932

This code adds indicators for timed four basic resistances, poison resistance, mana resistance and timed full ESP.
Indicators placed under location coordinates, left from the map on the main window.
Because of placement they are only visible if big_map option is turned on.

![image](https://user-images.githubusercontent.com/8179946/99272786-3efdcf00-2839-11eb-8a45-40f912a41f28.png)


Please leave comments if this code needs corrections to be accepted.

jezek commented:
I see, there is new code added to server, which sends new packets when drawing to client. Does the new server code works (have you tested it) with previous clients?

EvgenijK commented:
> I see, there is new code added to server, which sends new packets when drawing to client. Does the new server code works (have you tested it) with previous clients?

Thanks for pointing on this. I tested and it didn't work.

I updated game version and added client version check in server functions.

mhirki commented:
I think adding 7 new packet types is a total waste. Everything should be compressed into a single new packet with 32 bits reserved for flags.

EvgenijK commented:
> I think adding 7 new packet types is a total waste. Everything should be compressed into a single new packet with 32 bits reserved for flags.

I see two variants of this:
1) Send flags to toggle the indicators, i.e. if the indicator flag is in the packet, then toggle it on interface. This way requires to store flags on client.
2)  Send flags as state of indicators - if flag is set in the packet, then draw indicator, if flag isn't set, then erase it. There is no need to store info in this variant, but we have to send info about all indicators each time.

I hope i explained this variants clearly enough.
Personally I more like second variant.

What do you think, what would be better solution?

jezek commented:
@EvgenijK 
> What do you think, what would be better solution?

I would opt for the second variant too.

mhirki commented:
Yeah, variant 2 is defintely the answer.

EvgenijK commented:
@jezek @mhirki 
I made changes, waiting for the review.

jezek commented on src/server/xtra2.c:
```diff
@@ -616,6 +616,9 @@ bool set_tim_esp(int Ind, int v) {
 	p_ptr->update |= (PU_BONUS | PU_MONSTERS);
 
+    /* Redraw indicator */
+    p_ptr->redraw2 |= (PR2_INDICATORS);
```
Indentation mismatch?

EvgenijK replied:
fixed

jezek commented on src/server/xtra2.c:
```diff
@@ -2772,6 +2775,9 @@ bool set_oppose_acid(int Ind, int v) {
 	p_ptr->update |= (PU_BONUS);
 
+	/* Redraw indicator */
+    p_ptr->redraw2 |= (PR2_INDICATORS);
```
Indentation mismatch?

EvgenijK replied:
fixed

jezek commented on src/server/xtra2.c:
```diff
@@ -2818,7 +2824,10 @@ bool set_oppose_elec(int Ind, int v) {
 
-	/* Handle stuff */
+    /* Redraw indicator */
+    p_ptr->redraw2 |= (PR2_INDICATORS);
```
Indentation mismatch?

EvgenijK replied:
fixed

jezek commented on src/server/xtra2.c:
```diff
@@ -2864,6 +2873,9 @@ bool set_oppose_fire(int Ind, int v) {
 	p_ptr->update |= (PU_BONUS);
 
+    /* Redraw indicator */
+    p_ptr->redraw2 |= (PR2_INDICATORS);
```
Indentation mismatch?

EvgenijK replied:
fixed

jezek commented on src/server/xtra2.c:
```diff
@@ -2910,6 +2922,9 @@ bool set_oppose_cold(int Ind, int v) {
 	p_ptr->update |= (PU_BONUS);
 
+    /* Redraw indicator */
+    p_ptr->redraw2 |= (PR2_INDICATORS);
```
Indentation mismatch?

EvgenijK replied:
fixed

jezek commented on src/server/xtra2.c:
```diff
@@ -2956,6 +2971,9 @@ bool set_oppose_pois(int Ind, int v) {
 	p_ptr->update |= (PU_BONUS);
 
+    /* Redraw indicator */
+    p_ptr->redraw2 |= (PR2_INDICATORS);
```
Indentation mismatch?

jezek commented on src/server/xtra2.c:
```diff
@@ -3622,6 +3640,9 @@ bool do_divine_xtra_res(int Ind, int v) {
 	p_ptr->update |= (PU_BONUS);
 
+    /* Redraw indicator */
+    p_ptr->redraw2 |= (PR2_INDICATORS);
```
Indentation mismatch?

jezek commented on src/server/xtra1.c:
```diff
@@ -7327,6 +7366,11 @@ void redraw2_stuff(int Ind) {
 		prt_map(Ind, TRUE);
 	}
+
+    if (p_ptr->redraw2 & PR2_INDICATORS) {
```
Indentation mismatch?

jezek commented on src/client/c-xtra1.c:
```diff
@@ -1027,6 +1027,128 @@ void prt_bpr(byte bpr, byte attr) {
 }
 
+void prt_indicators(u32b indicators) {
+    prt_res_fire((indicators & IND_RES_FIRE) != 0);
```
Indentation mismatch?

jezek commented on src/client/nclient.c:
```diff
@@ -177,6 +177,8 @@ static void Receive_init(void) {
 
 	receive_tbl[PKT_GUIDE]		= Receive_Guide;
+
+    receive_tbl[PKT_INDICATORS] = Receive_indicators;
```
Indentation mismatch?

jezek commented on src/client/nclient.c:
```diff
@@ -4840,8 +4842,19 @@ int Receive_Guide(void) {
 }
 
+int Receive_indicators(void) {
+    int	n;
```
Indentation mismatch?

EvgenijK replied:
fixed

jezek commented on src/common/defines.h:
```diff
@@ -2206,6 +2206,28 @@
+#define ROW_RESIST_MANA		23
+#define COL_RESIST_MANA		6
+
+#define ROW_TEMP_ESP		25
```
Indentation consistency?

EvgenijK replied:
Tried to fixe this in CLion, Vim and github online editor, all of them shows that `25` on the same columt with other values

jezek replied:
That's wierd, but you're right. When I look at the [whole file](https://github.com/TomenetGame/tomenet/blob/e8fcb013e0297dad3f39141cddeb686559e299f3/src/common/defines.h#L2224-L2227), the indetation is ok.

jezek commented on src/server/nserver.c:
```diff
@@ -8850,6 +8850,19 @@ int Send_Guide(int Ind, byte search_type, int lineno, char* search_string) {
 }
 
+int Send_indicators(int Ind, u32b indicators) {
+    connection_t *connp = Conn[Players[Ind]->conn];
```
Indentation mismatch?

EvgenijK replied:
fixed

jezek commented on src/server/xtra1.c:
```diff
@@ -684,6 +684,42 @@ static void prt_extra_status(int Ind) {
 }
 
+static void prt_indicators(int Ind) {
+    player_type *p_ptr = Players[Ind];
```
Indentation mismatch?

EvgenijK replied:
fixed

jezek commented on src/common/pack.h:
```diff
@@ -253,6 +253,7 @@
 
 #define PKT_GUIDE		196	/* Server issues Guide-invocation on client-side */
 
+#define PKT_INDICATORS 197
```
I know, there are many misaligned constants, but try to align as the one before.

EvgenijK replied:
fixed

jezek reviewed (changes requested):
Apart from some indentation consistency mismatch, I see no logical flaws.
Note: I haven't tested if it compiles or behaves as expected (yet).

jezek commented on src/server/nserver.c:
```diff
@@ -8850,6 +8850,19 @@ int Send_Guide(int Ind, byte search_type, int lineno, char* search_string) {
+	if (!BIT(connp->state, CONN_PLAYING | CONN_READY)) {
+		errno = 0;
+		plog(format("Connection not ready for indicators (%d.%d.%d)",
+					Ind, connp->state, connp->id));
```
I would make this a oneliner. Other lines above are longer, so it wouldn't stand out and wrapping can do an editor.

jezek reviewed (changes requested):
Thank you for indentation change. Only the 2 line logging caught my eye. Sorry for missing it the first time and is just a suggestion of me. The maintainers have the last word, so let's wait for their comments. When I have more free time, I will try to compile (server, client) and test it,

jezek commented:
I've tested this PR. After merging to current mater, it compiles (server+client) on ubuntu 20.10 without introducing new warnings. I've also tested new (4.7.3.1) server against new (4.7.3.1) and old (4.7.3) client and vice versa. Everything works as described in this PR, no bugs spotted.

mhirki commented:
Thanks for the testing jezek. I did my own testing and everything seems to work. I'm gonna merge the code now.
